hashes/sha256: add iterative hmac calc. functions

hermannlelong opened the pull request:
Add init, update, and final functions to do HMAC
calculation using blocks, instead of one shot operation.

This is done the same way than for SHA256 iterative calculation.

The old function is still available.

Unit tests OK.

BytesGalore commented:
nice one. :+1:  I like your restructuring better than the former one-shot implementation. 
Could you provide an additional unittest that calls `hmac_sha256_update()` for updating several portions of a larger chunk successively?

I guess eventually we can just replace the former one-shot implementation with this one to provide a more uniform API.

hermannlelong commented:
Thanks for the remarks!

I've added the unittest requested, and I've also changed the `hmac_256` implementation to use the same code than for the iterative version.

vincent-d commented:
ping @BytesGalore

BytesGalore commented:
@hermannlelong, @vincent-d looks really good and works :+1:

vincent-d commented:
Squashed for @hermannlelong ;)

BytesGalore commented:
nice, lets see what our CIs say

BytesGalore commented:
and go
